core/src/local/persistent_state.rs

```diff
@@ -0,0 +1,891 @@
+            None => PersistentMeta::default(),
+        };
+
+        meta.epoch += 1;
+        let data = bincode::serialize(&meta).unwrap();
+        db.put(META_KEY, &data).unwrap();
+        meta
+    }
+
+    fn persist_meta(&mut self) {
+        // Stores the columns of self.indices in RocksDB so that we don't rebuild indices on recovery.
+        let indices: Vec<Vec<usize>> = self.indices.iter().map(|i| i.columns.clone()).collect();
+        let meta = PersistentMeta {
+            indices,
+            epoch: self.epoch,
+        };
+
+        let data = bincode::serialize(&meta).unwrap();
+        let db = self.db.as_ref().unwrap();
+        db.put(META_KEY, &data).unwrap();
+    }
+
+    fn get_index_id(&self, columns: &[usize]) -> Option<IndexID> {
+        self.indices
+            .iter()
+            .position(|index| &index.columns[..] == columns)
+            .and_then(|id| {
+                let id = id as IndexID;
+                // Increment the index ID with one if we have a primary key,
+                // as that'll always be index 0:
+                Some(match self.primary_key {
+                    Some(..) => id + 1,
+                    None => id,
+                })
+            })
+    }
+
+    // Selects a prefix of `key` without the epoch or sequence number.
+    //
+    // The RocksDB docs state the following:
+    // > If non-nullptr, use the specified function to determine the
+    // > prefixes for keys.  These prefixes will be placed in the filter.
+    // > Depending on the workload, this can reduce the number of read-IOP
+    // > cost for scans when a prefix is passed via ReadOptions to
+    // > db.NewIterator(). For prefix filtering to work properly,
+    // > "prefix_extractor" and "comparator" must be such that the following
+    // > properties hold:
+    //
+    // > 1) key.starts_with(prefix(key))
+    // > 2) Compare(prefix(key), key) <= 0.
+    // > 3) If Compare(k1, k2) <= 0, then Compare(prefix(k1), prefix(k2)) <= 0
+    // > 4) prefix(prefix(key)) == prefix(key)
+    //
+    // NOTE(ekmartin): Encoding the key size in the key increases the total size with 4 bytes.
+    // If we really wanted to avoid this while still maintaining the same serialization scheme
+    // we could do so by figuring out how many bytes our bincode serialized Vec<DataType> takes
+    // up here in transform_fn. Example:
+    // [DataType::Int(1), DataType::BigInt(10)] would be serialized as:
+    // 2u64 (vec length), 0u32 (enum variant), 1i32 (value), 1u32 (enum variant), 1i64 (value)
+    // By stepping through the serialized bytes and checking each enum variant we would know
+    // when we reached the end, and could then with certainty say whether we'd already
+    // prefix transformed this key before or not
+    // (without including the byte size of Vec<DataType>).
+    fn transform_fn(key: &[u8]) -> Vec<u8> {
+        if key == META_KEY {
+            return Vec::from(key);
```

Comment:
`unreachable!()` given `in_domain_fn`?

Reply:
One would think so, but it seems like RocksDB always uses slice transforms to create keys when you use a hash-based memtable factory (regardless of whether the domain function matches or not). The domain function then only applies to whether the slice transform is used to create bloom filters or not. 

Sort of described [here](https://github.com/facebook/rocksdb/blob/master/include/rocksdb/slice_transform.h#L42-L57), but still not too clear.

Added a comment about it in the `transform_fn` at least.